azure-kusto-data/azure/kusto/data/security.py

```diff
@@ -42,31 +76,34 @@ def __init__(self, kcsb):
             self._client_id = "db662dc1-0cfe-4e1c-a843-19a68e65be58"
             self._username = kcsb.aad_user_id
             self._password = kcsb.password
-            create_adal_context = True
         elif all([kcsb.application_client_id, kcsb.application_key]):
             self._authentication_method = AuthenticationMethod.aad_application_key
             self._client_id = kcsb.application_client_id
             self._client_secret = kcsb.application_key
-            create_adal_context = True
         elif all([kcsb.application_client_id, kcsb.application_certificate, kcsb.application_certificate_thumbprint]):
             self._authentication_method = AuthenticationMethod.aad_application_certificate
             self._client_id = kcsb.application_client_id
             self._certificate = kcsb.application_certificate
             self._thumbprint = kcsb.application_certificate_thumbprint
-            create_adal_context = True
         elif kcsb.msi_authentication:
             self._authentication_method = AuthenticationMethod.aad_msi
             self._msi_params = kcsb.msi_parameters
+            return
+        elif any([kcsb.user_token, kcsb.application_token]):
+            self._token = kcsb.user_token or kcsb.application_token
+            self._authentication_method = AuthenticationMethod.aad_token
+            return
+        elif kcsb.az_cli:
+            self._authentication_method = AuthenticationMethod.az_cli_profile
+            return
         else:
             self._authentication_method = AuthenticationMethod.aad_device_login
             self._client_id = "db662dc1-0cfe-4e1c-a843-19a68e65be58"
-            create_adal_context = True
 
-        if create_adal_context:
-            authority = kcsb.authority_id or "common"
-            aad_authority_uri = os.environ.get("AadAuthorityUri", "https://login.microsoftonline.com/")
-            full_authority_uri = aad_authority_uri + authority if aad_authority_uri.endswith("/") else aad_authority_uri + "/" + authority
-            self._adal_context = AuthenticationContext(full_authority_uri)
+        authority = kcsb.authority_id or "common"
```

Comment:
Following discussion, please postpone to all AuthenticationContext creation to acquire_authorization_header()